lib/server/server.mjs

```diff
@@ -0,0 +1,406 @@
+  /** @type {Map<string, RequestHandler[]>} */
+  #handlers = new Map()
+
+  /** @type {RequestExceptionHandler[]} */
+  #exceptionHandlers = [
+    defaultExceptionHandler
+  ]
+
+  /** @type {RequestHandler[]} */
+  #defaultHandlers = [
+    unknownSubjectHandler
+  ]
+
+  /** @type {Map<stream.Duplex, StreamContext>} */
+  #streams = new Map()
+
+  /** @type {pino.Logger} */
+  #logger
+
+  /**
+  * Construct a server.
+  *
+  * @param {object} [options] Options
+  * @param {pino.Logger} [options.logger] Logger; defaults to a `pino` logger
+  */
+  constructor (options) {
+    super()
+
+    this.#logger = options?.logger ?? pino({ name: 'nlon-server' })
+
+    this.on('error', err =>
+      this.#logger.error({ err }, 'Server stream error!'))
+  }
+
+  /**
+  * @summary Register a request handler.
+  *
+  * @description To process incoming messages, handlers are selected based on
+  * the message's `subject`. A single `subject` may map to multiple handlers -
+  * these are called in order of registration.
+  *
+  * The newly registered handlers are pushed to the end of the list, meaning that
+  * the most recently registered handler will be called last.
+  *
+  * @param {string} subject Subject
+  * @param {...RequestHandler} handlers Request handler
+  */
+  handle (subject, ...handlers) {
+    this.#ensureHandler(subject)
+    this.#handlers.get(subject).push(...handlers)
+  }
+
+  /**
+  * @summary Register a request exception handler.
+  *
+  * @description Exception handlers are called in case an exception is thrown
+  * during message processing. Multiple exception handlers may be registered,
+  * and will be called from **most recently registered** to **least recently
+  * registered.**
+  *
+  * This reverse order is implemented because a catch-all exception handler is
+  * always registered, which must be run last.
+  *
+  * @param {...RequestExceptionHandler} handlers Exception handlers
+  */
+  handleException (...handlers) {
+    this.#exceptionHandlers.push(...handlers)
+  }
+
+  /**
+  * @summary Set default handlers.
+  *
+  * @description Default handlers are called in case no request handlers are
+  * registered to the incoming message's `subject`. The processing works the
+  * same as regular request handlers - a set of handlers is ran until there's no
+  * more handlers left to call or the response is finished.
+  *
+  * This method **replaces** the list of default handlers, instead of appending
+  * to it.
+  *
+  * At least one handler must be present.
+  *
+  * By default, `unknownSubjectHandler` is registered as the default handler.
+  *
+  * @param {...RequestHandler} handlers Default handlers
+  */
+  defaultHandlers (...handlers) {
+    assert(handlers.length > 0, 'At least one default handler must be present!')
+
+    this.#defaultHandlers = handlers
+  }
+
+  /**
+  * @summary Configure the server.
+  *
+  * @description This method expects a method that can be called with the server
```

Comment:
I'd recommend adding some examples, quite frankly I don't really get it how I would need to use this method... :/